Fix misuse of SetHexExact

seelabs opened the pull request:
@rec @HowardHinnant Please confirm `SetHexExact` was misused and this fixes the issue.

rec commented on src/ripple/app/ledger/Ledger.cpp:
```diff
@@ -680,10 +680,14 @@ loadLedgerHelper(std::string const& sqlSuffix)
-    prevHash.SetHexExact (sPrevHash.value_or(""));
-    accountHash.SetHexExact (sAccountHash.value_or(""));
-    transHash.SetHexExact (sTransHash.value_or(""));
+    if (sLedgerHash)
```
Right, you don't need to zero 'em!

These four are obviously right.

nbougalis replied:
This relies on a design decision to have the default `base_uint` constructor zero-initialize. While we do this sort of thing all over the place, I'd rather we don't make a habit of it, but strive to reduce reliance on it.

HowardHinnant replied:
A free and concise way to address Nik's comment is:
````
uint256 prevHash{};
uint256 accountHash{};
// ...
````
This gives us leeway in the future to make `uint256` behave like `std::chrono::duration`:  It doesn't initialize for default-initialization and it zero-initializes when given braces (or parentheses) as above.

rec commented on src/ripple/app/ledger/Ledger.cpp:
```diff
@@ -835,7 +839,8 @@ Ledger::getHashesByIndex (std::uint32_t minSeq, std::uint32_t maxSeq)
-        hashes.second.SetHexExact (ph.value_or (""));
+        if (ph)
+            hashes.second.SetHexExact (*ph);
         if (!ph)
```
1. Should now be `else` on line 844.
2. Shouldn't you be emptying `hashes.second`?  It has a value of some type...

seelabs replied:
Oh, good catch on (2). Agree. Will fix (after lunch).

HowardHinnant commented:
I have no real objection to this P/R, but I'm having trouble understanding its motivation:
> `SetHexExact` may overwrite memory when given an empty string
Given "", `SetHexExact` is a no-op.  The replacement code given here is a no-op in exactly the same way.  What am I missing?

miguelportilla commented:
@seelabs Probably a good idea to fix this one too:
https://github.com/ripple/rippled/blob/develop/src/ripple/app/ledger/Ledger.cpp#L838

And I suggest adding an assert in SetHexExact. Something like:
assert(*(psz + sizeof(pn) * 2) == 0);

rec commented:
I think we discussed this on channel, but `SetHexExact` is under no circumstances a noop.

I actually think we should get rid of the `char const*` signature of `SetHexExact`, use the `std::string` one only, and always check the length - even in release, not just an `assert`, because otherwise we're getting UB.

If you look, the underlying class we're calling it with is always a `std::string` anyway...!  so why do we have the `char const*` overload?

HowardHinnant commented:
Just put it in a HelloWorld and test it out for yourself: http://melpon.org/wandbox/permlink/IjRmggjhWQmdOsBF
That being said, 3 members of my team find this code unreadable.  That is more than enough reason to change it.

HowardHinnant commented:
:+1:

seelabs commented:
@HowardHinnant I had misread the code - I missed the early exit from `charUnHex` when char is a non-hex character. I believe it is correct. However, I think I will keep this PR open, calling `setHashExact` with a null string is a no-op and this PR make that explicit - even if it doesn't fix a bug.

seelabs commented:
@HowardHinnant @rec I just pushed two new changes: 1) I renamed SetHexExact to SetHexUnchecked 2) I marked the LedgerClass as final (unrelated to SetHex, but I figured I'd sneak that small change into this PR). I also squashed, but there are so few changes in this PR I thought that would be OK.

HowardHinnant commented:
:+1:

rec commented:
I still think you should change `SetHexUnchecked` to take a string.  What's the point of converting all these strings to a `char const*` (which might involve a memory allocation!) and then passing them to a routine that takes its length??

HowardHinnant commented:
`string::c_str()` is marked `noexcept` and has a constant time complexity requirement.  This implies no underlying memory allocation, and no operation that is proportional to the length of the string.  As a practical matter, every implementation simply returns a pointer to the existing internal storage.  This is true both in the old reference counted implementations (which we still use in gcc-4.8), and the C++11/14 short-string optimization implementations (gcc-5.x, libc++, VS).

rec commented:
> As a practical matter, every implementation simply returns a pointer to the existing internal storage.

I thought that was true only of `std::string::data`?

This would imply that `std::string` always puts a closing `\0` on each string it contains.

Regardless, there's no advantage in calling `c_str()` and converting to `char const*` and then later calling  `strlen` when we are always starting with an `std::string` in the first place.

rec commented:
Ah, in C++11 and beyond, `std::data()` is now guaranteed to be the same as `std::c_str()`.  Good stuff!

HowardHinnant commented:
I'm lost.  Where is the call to `strlen`?

seelabs commented:
@miguelportilla The return tells the caller if the string was valid or not. Putting an `assert` in the code would trigger false positives.

seelabs commented:
@rec Given `c_str` doesn't allocate, I'd like to keep both the `const char*` interface and `std::string` interface.

rec commented:
But we don't actually _use_ the `char const*` interface - nor can I come up with a good reason why we would want to - and if we did, say for test code, it would still work with just the `std::string` overload.

rec commented:
Regardless, :+1:, but less is more.  :-)

nbougalis commented:
I don't, personally, like the renaming: `SetHexUnchecked` suggests that no checking is going on, which is not correct. There is error checking happening, and in fact, it's more strict than the checking in `SetHex`.

vinniefalco commented:
@nbougalis Please explain exactly what `SetHexExact` is supposed to do

nbougalis commented:
It's supposed to be used when you have a value that's exactly the correct number of bytes long, with no leading or trailing whitespace. Using it instead of `SetHex` allows you to have more robust error checking. Note, that you couldn't blindly replace it with `SetHex` anyways because it treats overlong values differently: `SetHex` happily accepts undersized inputs and happily discards the _leading_ bytes of overlong inputs.

My only objection is calling it `SetHexUnchecked` because it suggests that no checking is done - that you set the value to whatever is passed in and nothing else is done. But the function signature itself - returning a `bool` to indicate success or failure - rebuts that suggestion.

vinniefalco commented:
I see what you mean, yes you're right its not really an "unchecked" function.

seelabs commented:
Renamed `SetHexUnchecked` back to original `SetHexExact`.

rec commented:
Might I ask you to boot circleci (only)?

Otherwise, still :+1:.
